poseidon-sponge.md

```diff
@@ -0,0 +1,240 @@
+  - Else (there is room in `cache`):
+    1. Append `input` to the end of `cache`.
+
+- Else (`mode == squeezing`):
+  1. Drop any remaining cached output elements (by setting `cache` to `[]`).
+  2. Set `mode` to `absorbing`.
+  3. Set `cache` to `[input; 1]`.
+
+### Squeeze
+
+`PoseidonSponge.Squeeze` proceeds as follows:
+
+1. If `mode == squeezing` and `len(cache) == 0`:
+   1. Set `mode` to `absorbing`.
+   2. Set `cache` to `[]` (this is already the case, but for clarity this means there are
+      no elements to be absorbed).
+
+2. If `mode == absorbing`:
+   1. `new_output_elements = PoseidonSponge.PerformDuplex(cache)`.
+   2. Set `mode` to `squeezing`.
+   3. Set `cache` to `new_output_elements`.
+
+3. Remove the first element from `cache` and return it.
+
+## Hash functions
+
+### Merkle tree digest
+
+Merkle trees are built around `arity`-to-1 compression functions, that combine a set of
+nodes from one tree layer into a single node in the next layer.
+
+> TODO: Explain more clearly why the Merkle tree digest outputs `C` elements (as it's
+> confusing on its face that it becomes an `arity`-to-`C` compression function).
+
+`PoseidonHash.MerkleTree` takes as input:
+
+- The `arity` of the Merkle tree, of maximum value 32.
+  - The `RATE` of the sponge is set to `arity`.
+- A `PoseidonPermutation` with the desired properties; in particular, its width `T` must
+  equal `arity + C`, where `C` depends on the desired security level `M`.
+- `nodes`, an array of `arity` field elements.
+
+The digest proceeds as follows:
+
+```
+sponge = PoseidonSponge.Initialize(
+    PoseidonPermutation,
+    (1 << arity) - 1,
+    ZeroExtend,
+)
+
+for i in [0, arity):
+    sponge.Absorb(nodes[i])
+
+output = [0; C]
+for i in [0, C):
+    output[i] = sponge.Squeeze()
+return output
+```
+
+### Hash function with fixed-length input
+
+`PoseidonHash.ConstantLength` takes as input:
+
+- A `PoseidonPermutation` with the desired properties.
+- `message`, an array of `msg_len` field elements (where `msg_len` is a fixed constant in
+  some higher-level protocol).
+- An output size `out_len`.
+
+Hashing proceeds as follows:
+
+```
+sponge = PoseidonSponge.Initialize(
+    PoseidonPermutation,
+    (msg_len << 64) + (out_len - 1),
+    ZeroExtend,
+)
+
+for i in [0, msg_len):
+    sponge.Absorb(message[i])
+
+output = [0; out_len]
+for i in [0, out_len):
+    output[i] = sponge.Squeeze()
+return output
+```
+
+### Hash function with variable-length input
+
+`PoseidonHash.VariableLength` takes as input:
+
+- A `PoseidonPermutation` with the desired properties.
+- `message`, a sequence of field elements of variable length.
+- An output size `out_len`.
+
+Hashing proceeds as follows:
+
+```
+sponge = PoseidonSponge.Initialize(
+    PoseidonPermutation,
+    (1 << 64) + (out_len - 1),
+    ZeroExtend,
+)
+
+for i in [0, len(message)):
+    sponge.Absorb(message[i])
+
+# Domain-separate from fixed-length hash by terminating the message with a constant.
+sponge.Absorb(1)
```

Comment:
If the variable length is equal to the constant version's length, minus 1, then don't you get collisions still?

Reply:
there's the domain separation string that will prevent that :o